pyquil/latex/latex_generation.py

```diff
@@ -13,545 +13,478 @@
 #    limitations under the License.
 ##############################################################################
-# THIS FILE IS DERIVED AND MODIFIED FROM PROJECTQ. COPYRIGHT PROVIDED HERE:
+
+import sys
+from copy import copy
+from warnings import warn
+
+from pyquil import Program
+from pyquil.quil import Measurement, Gate, Pragma
+from pyquil.quilatom import format_parameter
+from pyquil.quilbase import AbstractInstruction
+from collections import defaultdict
+from typing import Optional
+
+if sys.version_info < (3, 7):
+    from pyquil.external.dataclasses import dataclass, replace
+else:
+    from dataclasses import dataclass, replace
+
+# Overview of LaTeX generation.
 #
-#   Copyright 2017 ProjectQ-Framework (www.projectq.ch)
+# The main entry point is the `to_latex` function below. Here are some high
+# points of the generation procedure:
 #
-#   Licensed under the Apache License, Version 2.0 (the "License");
-#   you may not use this file except in compliance with the License.
-#   You may obtain a copy of the License at
+# - The most basic building block are the TikZ operators, which are constructed
+#   by the functions below (e.g. TIKZ_CONTROL, TIKZ_NOP, TIKZ_MEASURE).
+# - TikZ operators are maintained by a DiagramState object, with roughly each
+#   qubit line in a diagram represented as a list of TikZ operators on the DiagramState.
+# - The DiagramBuilder is the actual driver. This traverses a Program and, for
+#   each instruction, performs a suitable manipulation of the DiagramState. At
+#   the end of this, the DiagramState is traversed and raw LaTeX is emitted.
+# - Most options are specified by DiagramSettings. One exception is this: it is possible
+#   to request that a certain subset of the program is rendered as a group (and colored
+#   as such). This is specified by a new pragma in the Program source:
 #
-#       http://www.apache.org/licenses/LICENSE-2.0
+#     PRAGMA LATEX_GATE_GROUP <name>?
+#     ...
+#     PRAGMA END_LATEX_GATE_GROUP
 #
-#   Unless required by applicable law or agreed to in writing, software
-#   distributed under the License is distributed on an "AS IS" BASIS,
-#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-#   See the License for the specific language governing permissions and
-#   limitations under the License.
-##############################################################################
+#   The <name> is optional, and will be used to label the group. Nested gate
+#   groups are currently not supported.
 
 
-from copy import copy
+@dataclass
+class DiagramSettings:
+    """
+    Settings to control the layout and rendering of circuits.
+    """
+
+    texify_numerical_constants: bool = True
+    """
+    Convert numerical constants, such as pi, to LaTeX form.
+    """
+
+    impute_missing_qubits: bool = False
+    """
+    Include qubits with indices between those explicitly referenced in the Quil program.
```

Comment:
like we discussed let's put all the Diagram and tikz stuff (so basically line 167 on) into a `_diagram.py` file and export just the `DiagramSettings` object

Reply:
Done